File: python3/max_number_of_k-sum_pairs.py
"""
Space   : O(n)
Time    : O(n)
"""


class Solution:
    def maxOperations(self, nums: List[int], k: int) -> int:
        d = {}
        ans = 0

        for x in nums:
            if x not in d:
                d[x] = 1
            else:
                d[x] += 1

        for key, val in d.items():
            if k - key in d:
                if k - key != key:
                    count = min(val, d[k - key])
                    d[k - key] -= count
                    d[key] -= count
                else:
                    count = val // 2
                    d[key] -= count * 2
                ans += count

        return ans

File: golang/max_number_of_k-sum_pairs.go
package main

// Space	: O(n)
// Time		: O(n)

func maxOperations(nums []int, k int) int {
	d := make(map[int]int)
	ans := 0

	for _, x := range nums {
		if _, ok := d[x]; ok {
			d[x]++
		} else {
			d[x] = 1
		}
	}

	for key, val := range d {
		if _, ok := d[k-key]; ok {
			count := 0
			if key*2 == k {
				count = val / 2
				d[key] %= 2
			} else {
				count = Min(val, d[k-key])
				d[key] -= count
				d[k-key] -= count
			}
			ans += count
		}
	}

	return ans
}
